### DISCUSSION TITLE
An Unofficial Stride Material System's "Effects and Shaders" Companion Guide

### DISCUSSION THREAD

**Discussion by Basewq** [stride3d/stride#2451] [General]
I've just released a repo with a guide to try help bridge the gap in knowledge from the official documentation [Effects and Shaders](https://doc.stride3d.net/latest/en/manual/graphics/effects-and-shaders/index.html) for Materials and working with the Material system in Stride, and a water shader demo project:
https://github.com/Basewq/BasewqStrideMaterialShaderGuide

Or just [click here to go straight to the guide page.](https://github.com/Basewq/BasewqStrideMaterialShaderGuide/blob/main/guide.md)

This is not meant for complete beginners, as that requires a guide on its own. This would be more targeted towards intermediate level programmer.

Here's a video of the water material in the demo project (not an artist so sorry that it isn't the world's greatest looking water!)

https://github.com/user-attachments/assets/6ed86fb5-36d0-4879-a044-73ffa332f553

Feel free to state what other gaps need filling (though keep in mind this more to do with the *Material* shader - things like compute shader is a different area), pass along your own tips and tricks here, or what you think should be added to the official documentation (don't say all of this).

**Comment by Eideren** [comment]
Did a really quick and dirty change to `MaterialWaveSurfaceNormal.sdsl`:
```cs
var worldPos = streams.PositionWS;
```
to
```cs
var worldPos = lerp(float3(streams.TexCoord.x, 0, streams.TexCoord.y)*100, streams.PositionWS, 0.75);
```
That way the gerstner expansion and contraction of vertices affects the microsurface of the water slightly;

https://github.com/user-attachments/assets/f3038bb3-d3fc-4a60-bac1-4de7d7ab6f12

The video starts with this change toggled on, then it is toggled off after a bit, then back on, etc.
It is fairly subtle, but it's a nice visual trick to make it seems as if the waves carry some amount of force behind them.

This is just a quick change though, I'm guessing moving this in the vertex shader/using world pos instead of the random `TexCoord` would be more appropriate

**Reply by Basewq** [reply]
Wow thanks, I didn't pick up on that. Updated the repo with this feature.
